File: app/src/main/java/com/protonvpn/android/utils/ServerManager.java
/*
 * Copyright (c) 2017 Proton Technologies AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.protonvpn.android.utils;

import android.content.Context;

import com.protonvpn.android.BuildConfig;
import com.protonvpn.android.models.config.UserData;
import com.protonvpn.android.models.profiles.Profile;
import com.protonvpn.android.models.profiles.SavedProfilesV3;
import com.protonvpn.android.models.profiles.ServerDeliver;
import com.protonvpn.android.models.profiles.ServerWrapper;
import com.protonvpn.android.models.vpn.Server;
import com.protonvpn.android.models.vpn.VpnCountry;

import org.jetbrains.annotations.TestOnly;
import org.joda.time.DateTime;
import org.joda.time.Minutes;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Random;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class ServerManager implements Serializable, ServerDeliver {

    private final List<VpnCountry> vpnCountries = new ArrayList<>();
    private final List<VpnCountry> secureCoreEntryCountries = new ArrayList<>();
    private final List<VpnCountry> secureCoreExitCountries = new ArrayList<>();
    private final SavedProfilesV3 savedProfiles;
    private DateTime updatedAt;

    private transient Context appContext;

    private transient LiveEvent updateEvent = new LiveEvent();
    private transient LiveEvent profilesUpdateEvent = new LiveEvent();

    private transient UserData userData;

    private ServerManager(Context appContext) {
        this.appContext = appContext.getApplicationContext();
        savedProfiles = Storage.load(SavedProfilesV3.class, SavedProfilesV3.defaultProfiles(appContext, this));
    }

    @Override
    public String toString() {
        return "vpnCountries: " + vpnCountries.size() + " entry: " + secureCoreEntryCountries.size()
            + " exit: " + secureCoreExitCountries.size() + " saved: " + (savedProfiles == null ?
            "null saved profiles" : (savedProfiles.getProfileList() == null ? "null profile list" :
            savedProfiles.getProfileList().size())) + " ServerManager Updated: " + updatedAt
            + " ";
    }

    @Inject
    public ServerManager(Context appContext, UserData userData) {
        this.appContext = appContext.getApplicationContext();
        ServerManager oldManager = Storage.load(ServerManager.class);
        if (oldManager != null) {
            vpnCountries.addAll(oldManager.getVpnCountries());
            secureCoreExitCountries.addAll(oldManager.getSecureCoreExitCountries());
            secureCoreEntryCountries.addAll(oldManager.getSecureCoreEntryCountries());
            updatedAt = oldManager.getUpdatedAt();
        }
        this.userData = userData;
        savedProfiles = Storage.load(SavedProfilesV3.class, SavedProfilesV3.defaultProfiles(appContext, this));
        reInitProfiles();
    }

    // TODO Remove this logic.
    // Whole profile providing should be moved to separate class outside of ServerManager
    private void reInitProfiles() {
        for (Profile profile : savedProfiles.getProfileList()) {
            profile.getServerWrapper().setDeliverer(this);
        }
        for (VpnCountry country : vpnCountries) {
            country.setDeliverer(this);
        }
        for (VpnCountry country : secureCoreExitCountries) {
            country.setDeliverer(this);
        }
        for (VpnCountry country : secureCoreEntryCountries) {
            country.setDeliverer(this);
        }
    }

    public LiveEvent getUpdateEvent() {
        return updateEvent;
    }

    public LiveEvent getProfilesUpdateEvent() {
        return profilesUpdateEvent;
    }

    public boolean isDownloadedAtLeastOnce() {
        return updatedAt != null && !vpnCountries.isEmpty();
    }

    public boolean isOutdated() {
        return updatedAt == null || vpnCountries.isEmpty()
            || Minutes.minutesBetween(updatedAt, new DateTime()).getMinutes() >= (BuildConfig.DEBUG ? 1 : 15);
    }

    public void clearCache() {
        updatedAt = null;
        Storage.delete(ServerManager.class);
    }

    public void setServers(List<Server> serverList) {
        vpnCountries.clear();
        secureCoreEntryCountries.clear();
        secureCoreExitCountries.clear();
        HashSet<String> countries = new HashSet<>();
        for (Server server : serverList) {
            countries.add(server.getFlag());
        }

        for (final String country : countries) {
            List<Server> servers = filterServers(serverList,
                server -> !server.isSecureCoreServer() && server.getFlag().equals(country));
            VpnCountry vpnCountry = new VpnCountry(country, servers, this, getBestScoreServer(servers));
            vpnCountries.add(vpnCountry);
        }

        for (final String country : countries) {
            if (country.equals("IS") || country.equals("SE") || country.equals("CH")) {
                List<Server> servers = filterServers(serverList,
                    server -> server.isSecureCoreServer() && server.getEntryCountry()
                        .equalsIgnoreCase(country));

                VpnCountry vpnCountry = new VpnCountry(country, servers, this, getBestScoreServer(servers));
                vpnCountry.addBestConnectionToList(false);
                secureCoreEntryCountries.add(vpnCountry);
            }
        }

        for (final String country : countries) {
            List<Server> servers = filterServers(serverList,
                server -> server.isSecureCoreServer() && server.getExitCountry().equalsIgnoreCase(country));
            VpnCountry vpnCountry = new VpnCountry(country, servers, this, getBestScoreServer(servers));
            if (!vpnCountry.getServerList().isEmpty()) {
                vpnCountry.addBestConnectionToList(false);
                secureCoreExitCountries.add(vpnCountry);
            }
        }

        sortVpnCountries(vpnCountries);
        sortVpnCountries(secureCoreExitCountries);

        updatedAt = new DateTime();
        Storage.save(this);
        updateEvent.emit();
        profilesUpdateEvent.emit();
    }

    private void sortVpnCountries(List<VpnCountry> list) {
        Collections.sort(list, (lhs, rhs) -> {
            if (userData.isFreeUser()) {
                if (lhs.hasAccessibleServer(userData)) {
                    return rhs.hasAccessibleServer(userData) ?
                        rhs.getCountryName().compareTo(lhs.getCountryName()) : -1;
                }
                if (rhs.hasAccessibleServer(userData)) {
                    return 1;
                }
                return lhs.getCountryName().compareTo(rhs.getCountryName());
            }
            return lhs.getCountryName().compareTo(rhs.getCountryName());
        });
    }

    public List<VpnCountry> getVpnCountries() {
        return vpnCountries;
    }

    @Nullable
    public Profile getDefaultConnection() {
        Profile profile = userData.getDefaultConnection() == null ? getSavedProfiles().get(0) :
            userData.getDefaultConnection();
        profile.getServerWrapper().setDeliverer(this);
        return profile;
    }

    public List<VpnCountry> getSecureCoreEntryCountries() {
        return secureCoreEntryCountries;
    }

    private List<Server> filterServers(List<Server> serverList, CollectionTools.Predicate<Server> predicate) {
        List<Server> filtered = new ArrayList<>();
        for (Server server : serverList) {
            if (predicate.contains(server)) {
                filtered.add(server);
            }
        }
        return filtered;
    }

    @Nullable
    public Server getServerById(String serverId) {
        for (VpnCountry vpnCountry : getVpnCountries()) {
            for (Server server : vpnCountry.getServerList()) {
                if (server.getServerId().equals(serverId)) {
                    return server;
                }
            }
        }
        for (VpnCountry vpnCountry : getSecureCoreEntryCountries()) {
            for (Server server : vpnCountry.getServerList()) {
                if (server.getServerId().equals(serverId)) {
                    return server;
                }
            }
        }
        return null;
    }

    @Nullable
    public VpnCountry getVpnCountry(String country, boolean secureCoreCountry) {
        for (VpnCountry vpnCountry : secureCoreCountry ? getSecureCoreEntryCountries() : getVpnCountries()) {
            if (vpnCountry.getFlag().equals(country)) {
                return vpnCountry;
            }
        }
        return null;
    }

    @Nullable
    public VpnCountry getVpnExitCountry(String country, boolean secureCoreCountry) {
        for (VpnCountry vpnCountry : secureCoreCountry ? getSecureCoreExitCountries() : getVpnCountries()) {
            if (vpnCountry.getFlag().equals(country)) {
                return vpnCountry;
            }
        }
        return null;
    }

    public Server getBestScoreServer(VpnCountry country) {
        return getBestScoreServer(country.getConnectableServers());
    }

    @Nullable
    public Server getBestScoreServer(List<Server> serverList) {
        Server bestScore = null;
        boolean hasAccessAtAll = userData.hasAccessToAnyServer(serverList);
        for (Server server : serverList) {
            if (!server.getKeywords().contains("tor") && server.isOnline()) {
                if (bestScore == null) {
                    bestScore = server;
                }
                else if (
                    (bestScore.getScore() > server.getScore() || !userData.hasAccessToServer(bestScore)) && (
                        !hasAccessAtAll || userData.hasAccessToServer(server))
                        || server.isSecureCoreServer()) {
                    bestScore = server;
                }
            }
        }
        return bestScore;
    }

    @Nullable
    public Server getRandomServer() {
        List<VpnCountry> list =
            userData.isSecureCoreEnabled() ? getSecureCoreExitCountries() : getVpnCountries();
        if (!userData.isSecureCoreEnabled()) {
            list = CollectionTools.findAll(list, item -> item.hasAccessibleServer(userData));
        }
        return list.isEmpty() ? null : getRandomServerForCountry(list.get(new Random().nextInt(list.size())));
    }

    @NonNull
    public Server getRandomServerForCountry(VpnCountry country) {
        List<Server> serverList = country.getServerList();
        if (!userData.isSecureCoreEnabled()) {
            serverList = CollectionTools.findAll(serverList, userData::hasAccessToServer);
        }
        if (serverList.size() == 0) {
            serverList = country.getServerList();
        }
        return serverList.get(new Random().nextInt(serverList.size()));
    }

    public Server getRandomServerForCountryExcluded(VpnCountry country, Server excludedServer) {
        List<Server> serverList = country.getServerList();
        if (!userData.isSecureCoreEnabled()) {
            serverList = CollectionTools.findAll(serverList, item -> userData.hasAccessToServer(item) && !item
                .getServerName()
                .equals(excludedServer.getServerName()));
        }
        if (serverList.size() == 0) {
            serverList = country.getServerList();
        }
        return serverList.get(new Random().nextInt(serverList.size()));
    }

    public Server getBestScoreServerFromAll() {
        Server bestScore = null;
        for (VpnCountry country : userData.isSecureCoreEnabled() ? getSecureCoreExitCountries() :
            getVpnCountries()) {
            if (country.hasAccessibleServer(userData) || userData.isSecureCoreEnabled()) {
                Server countryBestServer = getBestScoreServer(country);
                if (bestScore == null) {
                    bestScore = countryBestServer;
                }
                else if (countryBestServer != null && (bestScore.getScore() > countryBestServer.getScore())) {
                    bestScore = countryBestServer;
                }
            }
        }
        return bestScore;
    }

    public List<Profile> getSavedProfiles() {
        return savedProfiles.getProfileList();
    }

    public void deleteSavedProfiles() {
        List<Profile> defaultProfiles = SavedProfilesV3.defaultProfiles(appContext, this).getProfileList();
        for (Profile profile : getSavedProfiles()) {
            if (!defaultProfiles.contains(profile)) {
                deleteProfile(profile);
            }
        }
    }

    public void addToProfileList(String serverName, String color, Server server) {
        Profile newProfile = new Profile(serverName, color, ServerWrapper.makeWithServer(server, this));
        newProfile.getServerWrapper().setSecureCoreCountry(userData.isSecureCoreEnabled());
        addToProfileList(newProfile);
    }

    public boolean addToProfileList(Profile profileToSave) {
        if (!savedProfiles.getProfileList().contains(profileToSave)) {
            savedProfiles.getProfileList().add(profileToSave);
            Storage.save(savedProfiles);
            profilesUpdateEvent.emit();
            return true;
        }
        return false;
    }

    public void editProfile(Profile oldProfile, Profile profileToSave) {
        if (oldProfile.equals(getDefaultConnection())) {
            userData.setDefaultConnection(profileToSave);
        }
        savedProfiles.getProfileList().set(savedProfiles.getProfileList().indexOf(oldProfile), profileToSave);
        Storage.save(savedProfiles);
        profilesUpdateEvent.emit();
    }

    public void deleteProfile(Profile profileToSave) {
        savedProfiles.getProfileList().remove(profileToSave);
        Storage.save(savedProfiles);
        profilesUpdateEvent.emit();
    }

    public List<VpnCountry> getSecureCoreExitCountries() {
        return secureCoreExitCountries;
    }

    public DateTime getUpdatedAt() {
        return updatedAt;
    }

    @Nullable
    public Server getServerFromWrap(ServerWrapper wrapper) {
        switch (wrapper.type) {
            case FASTEST:
                return getBestScoreServerFromAll();
            case RANDOM:
                return getRandomServer();
            case RANDOM_IN_COUNTRY:
                VpnCountry country = wrapper.secureCoreCountry ? getVpnExitCountry(wrapper.country, true) :
                    getVpnCountry(wrapper.country, false);
                return country != null ? getRandomServerForCountry(country) : null;
            case FASTEST_IN_COUNTRY:
                VpnCountry vpnCountry = wrapper.secureCoreCountry ? getVpnExitCountry(wrapper.country, true) :
                    getVpnCountry(wrapper.country, false);
                return vpnCountry != null ? getBestScoreServer(vpnCountry) : null;
            case DIRECT:
                return getServerById(wrapper.serverId);
        }
        throw new RuntimeException("Incorrect server type in profile");
    }

    @Nullable
    @Override
    public Server getServer(ServerWrapper wrapper) {
        return getServerFromWrap(wrapper);
    }

    @Override
    public boolean hasAccessToServer(Server server) {
        return userData.hasAccessToServer(server);
    }

    @TestOnly
    public VpnCountry getFirstNotAccessibleVpnCountry() {
        List<VpnCountry> countries = getVpnCountries();

        for (VpnCountry country : countries) {
            if (!country.hasAccessibleServer(userData)) {
                return country;
            }
        }
        throw new UnsupportedOperationException("Should only use this method on free tiers");
    }
}

File: app/src/main/java/com/protonvpn/android/utils/ServerManager.kt
/*
 * Copyright (c) 2017 Proton AG
 *
 * This file is part of ProtonVPN.
 *
 * ProtonVPN is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProtonVPN is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ProtonVPN.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.protonvpn.android.utils

import androidx.annotation.VisibleForTesting
import com.protonvpn.android.BuildConfig
import com.protonvpn.android.appconfig.UserCountryPhysical
import com.protonvpn.android.auth.data.VpnUser
import com.protonvpn.android.auth.data.hasAccessToServer
import com.protonvpn.android.excludedlocations.ExcludedLocations
import com.protonvpn.android.logging.ProtonLogger
import com.protonvpn.android.models.profiles.Profile
import com.protonvpn.android.models.profiles.ServerWrapper.ProfileType
import com.protonvpn.android.models.vpn.GatewayGroup
import com.protonvpn.android.models.vpn.VpnCountry
import com.protonvpn.android.models.vpn.usecase.SmartProtocols
import com.protonvpn.android.models.vpn.usecase.supportsProtocol
import com.protonvpn.android.redesign.CountryId
import com.protonvpn.android.redesign.vpn.AnyConnectIntent
import com.protonvpn.android.redesign.vpn.ConnectIntent
import com.protonvpn.android.redesign.vpn.ServerFeature
import com.protonvpn.android.redesign.vpn.isExcluded
import com.protonvpn.android.redesign.vpn.satisfiesFeatures
import com.protonvpn.android.servers.Server
import com.protonvpn.android.servers.ServersDataManager
import com.protonvpn.android.servers.ServersDataManager.ServerLists
import com.protonvpn.android.servers.ServersResult
import com.protonvpn.android.servers.api.ConnectingDomain
import com.protonvpn.android.servers.api.LogicalsStatusId
import com.protonvpn.android.vpn.ProtocolSelection
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import kotlinx.coroutines.yield
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class ServerManagerState(
    val serverListAppVersionCode: Int = 0,
    @Deprecated("The timestamp is now stored by ServersDataManager.")
    val lastUpdateTimestamp: Long = 0L,

    /** Can be checked even before servers are loaded from storage */
    val hasGateways: Boolean = false,
    val hasCountries: Boolean = false,
)

@Deprecated("Use ServerManager2 in new code")
@Singleton
class ServerManager @Inject constructor(
    mainScope: CoroutineScope,
    val serversDataManager: ServersDataManager,
    val physicalUserCountry: UserCountryPhysical,
) {
    private var savedState: ServerManagerState

    private var guestHoleServers: List<Server>? = null
    private val isLoaded = MutableStateFlow(false)

    // Expose a version number of the server list so that it can be used in flow operators like
    // combine to react to updates.
    val serverListVersion = MutableStateFlow(0)

    val isDownloadedAtLeastOnce get() = serversDataManager.lastUpdateTimestamp > 0
    /** Can be checked even before servers are loaded from storage */
    val hasCountriesFlow = serverListVersion.map { savedState.hasCountries }.distinctUntilChanged()
    val hasGatewaysFlow = serverListVersion.map { savedState.hasGateways }.distinctUntilChanged()

    // The cached values are to be used only by legacy code.
    private val serversDataCachedFlow =
        serversDataManager.serverLists.stateIn(mainScope, SharingStarted.Eagerly, ServerLists.Empty)
    private val serversDataCached get() = serversDataCachedFlow.value // May be empty.

    suspend fun needsUpdate(): Boolean {
        val serversData = serversDataManager.serverLists.first()
        return serversDataManager.lastUpdateTimestamp == 0L || serversData.allServers.isEmpty() ||
            !haveWireGuardSupport() || savedState.serverListAppVersionCode < BuildConfig.VERSION_CODE
    }

    val allServers get() = serversDataCached.allServers
    val allServersByScore get() = serversDataCached.allServersByScore

    val freeCountries
        get() = getVpnCountries()
            .filter { country -> country.serverList.any { server -> server.isFreeServer } }

    init {
        val loadedState = Storage.load(ServerManager::class.java, ServerManagerState.serializer())
        if (loadedState != null) {
            savedState = loadedState
            serverListVersion.value = 1
        } else {
            savedState = ServerManagerState()
        }

        mainScope.launch(start = CoroutineStart.UNDISPATCHED) {
            // Starts undispatched to call load() immediately. This makes sure initialization
            // happens before any other calls to ServersDataManager.
            // It's not very robust, we should improve it in the future.
            val loaded = serversDataManager.load()
            if (!loaded) {
                // We had servers saved but failed to load them, reset state.
                updateAndSave(
                    savedState.copy(
                        lastUpdateTimestamp = 0L,
                        hasGateways = false,
                        hasCountries = false,
                    )
                )
            } else if (
                serversDataManager.lastUpdateTimestamp == 0L &&
                savedState.lastUpdateTimestamp > 0
            ) {
                serversDataManager.updateLastUpdateTimestamp(savedState.lastUpdateTimestamp)
                updateAndSave(savedState.copy(lastUpdateTimestamp = 0L))
            }
            // Yield to allow pending work to process, most importantly serversDataCachedFlow.
            // This is a hack, once we rewrite the TV UI all the synchronous getters will be
            // removed and this will not be needed.
            yield()
            // Notify of loaded state and update after everything has been updated.
            isLoaded.value = true
        }
        serversDataManager.serverLists.onEach { serversData ->
            updateAndSave(
                savedState.copy(
                    serverListAppVersionCode = BuildConfig.VERSION_CODE,
                    hasGateways = serversData.gateways.isNotEmpty(),
                    hasCountries = serversData.vpnCountries.isNotEmpty(),
                )
            )
            serverListVersion.value++
        }.launchIn(mainScope)
    }

    suspend fun ensureLoaded() {
        isLoaded.first { isLoaded -> isLoaded }
    }

    fun getExitCountries(secureCore: Boolean) = if (secureCore)
        serversDataCached.secureCoreExitCountries else serversDataCached.vpnCountries

    override fun toString(): String {
        val lastUpdateTimestampLog = serversDataManager.lastUpdateTimestamp
                .takeIf { it != 0L }
                ?.let { ProtonLogger.formatTime(it) }
        return "vpnCountries: ${serversDataCached.vpnCountries.size} gateways: ${serversDataCached.gateways.size}" +
            " exit: ${serversDataCached.secureCoreExitCountries.size} " +
            "ServerManager Updated: $lastUpdateTimestampLog"
    }

    fun clearCache() {
        savedState = ServerManagerState()
        Storage.delete(ServerManager::class.java)
        // The server list itself is not deleted.
    }

    suspend fun setGuestHoleServers(serverList: List<Server>) {
        DebugUtils.debugAssert("Guest hole servers can only be set when regular servers are not available") {
            !isDownloadedAtLeastOnce
        }
        setServers(serverList, null)
    }

    fun getDownloadedServersForGuestHole(
        serverCount: Int,
        protocol: ProtocolSelection,
        smartProtocols: SmartProtocols
    ): List<Server> {
        val bestScoreServer = getBestScoreServer(
            allServersByScore.filter { it.online },
            vpnUser = null,
            protocol,
            smartProtocols
        )
        val servers = listOfNotNull(bestScoreServer) +
            getExitCountries(false).flatMap { country ->
                country.serverList.filter { it.online && supportsProtocol(it, protocol, smartProtocols) }
            }

        return servers.takeRandomStable(serverCount).shuffled().distinct()
    }

    @VisibleForTesting
    suspend fun setServers(
        serverList: List<Server>,
        statusId: LogicalsStatusId?,
        retainIDs: Set<String> = emptySet()
    ) {
        ensureLoaded()
        serversDataManager.replaceServers(serverList, statusId, retainIDs)
    }

    suspend fun updateServerDomainStatus(connectingDomain: ConnectingDomain) {
        ensureLoaded()
        serversDataManager.updateServerDomainStatus(connectingDomain)
    }

    fun getServerById(id: String) =
        allServers.firstOrNull { it.serverId == id } ?: guestHoleServers?.firstOrNull { it.serverId == id }

    fun getVpnCountries(): List<VpnCountry> = serversDataCached.vpnCountries.sortedByLocaleAware { it.countryName }

    fun getGateways(): List<GatewayGroup> = serversDataCached.gateways

    @Deprecated("Use the suspending getVpnExitCountry from ServerManager2")
    fun getVpnExitCountry(countryCode: String, secureCoreCountry: Boolean): VpnCountry? =
        getExitCountries(secureCoreCountry).firstOrNull { it.flag == countryCode }

    @VisibleForTesting
    fun getBestScoreServer(
        serverList: Iterable<Server>,
        vpnUser: VpnUser?,
        protocol: ProtocolSelection,
        smartProtocols: SmartProtocols
    ): Server? {
        val eligibleServers = serverList.sortedBy { it.score }.asSequence()
            .filter { supportsProtocol(it, protocol, smartProtocols) }
        return with(eligibleServers) {
            firstOrNull { vpnUser.hasAccessToServer(it) && it.online }
                ?: firstOrNull { vpnUser.hasAccessToServer(it) }
                ?: firstOrNull()
        }
    }

    fun getRandomServer(vpnUser: VpnUser?, protocol: ProtocolSelection, smartProtocols: SmartProtocols): Server? {
        val allCountries = getExitCountries(secureCore = false)
        val accessibleCountries = allCountries.filter { it.hasAccessibleOnlineServer(vpnUser) }
        return accessibleCountries
            .ifEmpty { allCountries }
            .randomNullable()
            ?.let { getRandomServer(it, vpnUser, protocol, smartProtocols) }
    }

    private fun getRandomServer(
        country: VpnCountry,
        vpnUser: VpnUser?,
        protocol: ProtocolSelection,
        smartProtocols: SmartProtocols
    ): Server? {
        val online = country.serverList.filter(Server::online)
        val accessible = online.filter {
            vpnUser.hasAccessToServer(it) && supportsProtocol(it, protocol, smartProtocols)
        }
        return accessible.randomNullable()
    }

    fun getSecureCoreExitCountries(): List<VpnCountry> =
        serversDataCached.secureCoreExitCountries.sortedByLocaleAware { it.countryName }

    @Deprecated("Use getServerForConnectIntent")
    fun getServerForProfile(
        profile: Profile,
        vpnUser: VpnUser?,
        protocol: ProtocolSelection,
        smartProtocols: SmartProtocols
    ): Server? {
        val wrapper = profile.wrapper
        return when (wrapper.type) {
            ProfileType.FASTEST -> {
                val tvServers = allServersByScore.filter { it.online && !it.isGatewayServer }
                getBestScoreServer(tvServers, vpnUser, protocol, smartProtocols)
            }

            ProfileType.RANDOM ->
                getRandomServer(vpnUser, protocol, smartProtocols)

            ProfileType.RANDOM_IN_COUNTRY ->
                getVpnExitCountry(wrapper.country, false)?.let {
                    getRandomServer(it, vpnUser, protocol, smartProtocols)
                }

            ProfileType.FASTEST_IN_COUNTRY ->
                getVpnExitCountry(wrapper.country, false)?.let {
                    getBestScoreServer(it.serverList, vpnUser, protocol, smartProtocols)
                }

            ProfileType.DIRECT ->
                getServerById(wrapper.serverId!!)
        }
    }

    fun getBestServerForConnectIntent(
        connectIntent: AnyConnectIntent,
        vpnUser: VpnUser?,
        protocol: ProtocolSelection,
        smartProtocols: SmartProtocols,
        excludedLocations: ExcludedLocations,
    ): Server? = forConnectIntent(
        connectIntent = connectIntent,
        fallbackResult = null,
        excludedLocations = excludedLocations,
    ) { serversResult ->
        getBestScoreServer(serversResult.servers, vpnUser, protocol, smartProtocols)
    }

    /*
     * Perform operations related to ConnectIntent.
     *
     * ConnectIntent can specify either a fastest server overall, fastest in country, a specific server and so on.
     * Use this function to implement operations for a ConnectIntent like checking if its country/city/server is
     * available.
     */
    fun <T> forConnectIntent(
        connectIntent: AnyConnectIntent,
        fallbackResult: T,
        excludedLocations: ExcludedLocations,
        onServersResult: (ServersResult) -> T,
    ): T {
        fun Iterable<Server>.filterFeatures() = filter { it.satisfiesFeatures(connectIntent.features) }
        fun Server.satisfiesFeatures() = satisfiesFeatures(connectIntent.features)

        fun allRegularServersFor(
            secureCore: Boolean,
            features: Set<ServerFeature>,
            excludedCountryId: CountryId?,
        ): Sequence<Server> {
            var serversSequence = allServersByScore.asSequence()

            if (excludedCountryId != null) {
                serversSequence = serversSequence.filterNot { it.exitCountry == excludedCountryId.countryCode }
            }

            return serversSequence.filter { server ->
                server.isSecureCoreServer == secureCore &&
                        !server.isGatewayServer &&
                        server.satisfiesFeatures(features)
            }
        }

        return when (connectIntent) {
            is ConnectIntent.FastestInCountry ->
                if (connectIntent.country.isFastest) {
                    var hasAppliedExclusions = false

                    allRegularServersFor(
                        secureCore = false,
                        features = connectIntent.features,
                        excludedCountryId = ifOrNull(
                            predicate = connectIntent.country.isFastestExcludingMyCountry,
                            block = physicalUserCountry::invoke,
                        ),
                    )
                        .asIterable()
                        .filterNot { server ->
                            server.isExcluded(excludedLocations).also { isExcluded ->
                                if (isExcluded) {
                                    hasAppliedExclusions = true
                                }
                            }
                        }
                        .let { servers -> ServersResult.Regular(servers, hasAppliedExclusions) }
                        .let { serversResult -> onServersResult(serversResult) }
                } else {
                    getVpnExitCountry(connectIntent.country.countryCode, false)
                        ?.serverList
                        ?.filterFeatures()
                        .handleServersResult(onServersResult, fallbackResult)
                }

            is ConnectIntent.FastestInCity -> {
                getVpnExitCountry(connectIntent.country.countryCode, false)
                    ?.serverList
                    ?.filter { server -> server.city == connectIntent.cityEn && server.satisfiesFeatures() }
                    .handleServersResult(onServersResult, fallbackResult)
            }

            is ConnectIntent.FastestInState -> {
                getVpnExitCountry(connectIntent.country.countryCode, false)
                    ?.serverList
                    ?.filter { server -> server.state == connectIntent.stateEn && server.satisfiesFeatures() }
                    .handleServersResult(onServersResult, fallbackResult)
            }

            is ConnectIntent.SecureCore ->
                if (connectIntent.exitCountry.isFastest) {
                    var hasAppliedExclusions = false

                    allRegularServersFor(
                        secureCore = true,
                        features = connectIntent.features,
                        excludedCountryId = ifOrNull(
                            predicate = connectIntent.exitCountry.isFastestExcludingMyCountry,
                            block = physicalUserCountry::invoke,
                        ),
                    )
                        .asIterable()
                        .filterNot { server ->
                            server.isExcluded(excludedLocations).also { isExcluded ->
                                if (isExcluded) {
                                    hasAppliedExclusions = true
                                }
                            }
                        }
                        .let { servers -> ServersResult.Regular(servers, hasAppliedExclusions) }
                        .let { serversResult -> onServersResult(serversResult) }
                } else {
                    getVpnExitCountry(connectIntent.exitCountry.countryCode, true)
                        ?.serverList
                        ?.let { servers ->
                            if (connectIntent.entryCountry.isFastest) {
                                servers.filterFeatures()
                            } else {
                                servers.find { server ->
                                    server.entryCountry == connectIntent.entryCountry.countryCode && server.satisfiesFeatures()
                                }
                                ?.let(::listOf)
                            }
                        }
                        .handleServersResult(onServersResult, fallbackResult)
                }

            is ConnectIntent.Gateway -> {
                if (connectIntent.serverId != null) {
                    getServerById(connectIntent.serverId).handleServersResult(onServersResult, fallbackResult)
                } else {
                    getGateways()
                        .find { it.name() == connectIntent.gatewayName }
                        ?.serverList
                        ?.filterFeatures()
                        .handleServersResult(onServersResult, fallbackResult)
                }
            }

            is ConnectIntent.Server -> {
                getServerById(connectIntent.serverId).handleServersResult(onServersResult, fallbackResult)
            }

            is AnyConnectIntent.GuestHole -> {
                getServerById(connectIntent.serverId).handleServersResult(onServersResult, fallbackResult)
            }
        }
    }

    private fun updateAndSave(newState: ServerManagerState) {
        savedState = newState
        Storage.save(savedState, ServerManager::class.java, ServerManagerState.serializer())
    }

    private fun haveWireGuardSupport() =
        serversDataCached.allServers.any { server -> server.connectingDomains.any { it.publicKeyX25519 != null } }

    private fun <T> Server?.handleServersResult(
        onServersResult: (ServersResult) -> T,
        fallbackResult: T,
    ): T = this?.let(::listOf).handleServersResult(onServersResult, fallbackResult)

    private fun <T> List<Server>?.handleServersResult(
        onServersResult: (ServersResult) -> T,
        fallbackResult: T,
    ): T = this?.let { servers -> onServersResult(ServersResult.Regular(servers)) } ?: fallbackResult

}
